### PR TITLE
[Merged by Bors] - Smooth Transition between Animations

### PR DESCRIPTION
**Pull request by smessmer**
# Objective

- Fixes https://github.com/bevyengine/bevy/discussions/6338

This PR allows for smooth transitions between different animations.

## Solution

- This PR uses very simple linear blending of animations.
- When starting a new animation, you can give it a duration, and throughout that duration, the previous and the new animation are being linearly blended, until only the new animation is running.
- I'm aware of https://github.com/bevyengine/rfcs/pull/49 and https://github.com/bevyengine/rfcs/pull/51, which are more complete solutions to this problem, but they seem still far from being implemented. Until they're ready, this PR allows for the most basic use case of blending, i.e. smoothly transitioning between different animations.

## Migration Guide

- no bc breaking changes

### CODE REVIEW

File: examples/animation/animated_transform.rs
Comment (DevinLeamy) [inline examples/animation/animated_transform.rs:117]: Could you add an example that uses a transition duration?

File: crates/bevy_animation/src/lib.rs
Comment (smessmer) [reply crates/bevy_animation/src/lib.rs:180]: `self.start()` will reset the player, i.e. it will run the new animation even if the previous one was paused. So it should already resume a paused animation. Or did I misunderstand you?

File: crates/bevy_animation/src/lib.rs
Comment (smessmer) [reply crates/bevy_animation/src/lib.rs:330]: This is pre-PR code. GitHub doesn't show the changed lines very well. Not entirely sure why it is there but didn't want to touch it since there might be a corner case I missed.

Comment (mockersf) [reply crates/bevy_animation/src/lib.rs:330]: because animation can be played in reverse

File: crates/bevy_animation/src/lib.rs
Comment (james7132) [inline crates/bevy_animation/src/lib.rs:440]:
```suggestion
fn update_transitions(player: &mut AnimationPlayer, time: &Time) {
```

We don't need to prefix our private module functions with `_`.

File: crates/bevy_animation/src/lib.rs
Comment (james7132) [inline crates/bevy_animation/src/lib.rs:295]: This redoes the hierarchy query for the same bones repeatedly, which can be the most expensive part of applying animations as it traverses the hierarchy and does name comparisons at every level. Can we move the animation sample to the inner part of the loop instead of into the outside?

**Review by james7132 — CHANGES_REQUESTED** [review]
Generally looks good. Lines up decently well with the existing RFC designs, though there is a major perf question to be addressed.

File: crates/bevy_animation/src/lib.rs
Comment (DevinLeamy) [reply crates/bevy_animation/src/lib.rs:180]: Ah, that makes sense.

File: examples/animation/animated_transform.rs
Comment (smessmer) [reply examples/animation/animated_transform.rs:117]: Ok I added it to the `animated_fox` example

File: crates/bevy_animation/src/lib.rs
Comment (smessmer) [reply crates/bevy_animation/src/lib.rs]: Hm good point. It's not very straightforward though because different animations may attach to different bones, so I cannot just make the bones the outer loop and the animations the inner loop. I could first calculate a union set of bones over all animations and then iterate over that in the outer loop? That would add some overhead for the calculation of the union set though. What do you think?

Comment (smessmer) [reply crates/bevy_animation/src/lib.rs]: We could also store an `Entity` id instead of an `EntityPath` in `AnimationClip`. `Entity` id shouldn't change after things are instantiated. Or would that break something?

Comment (james7132) [reply crates/bevy_animation/src/lib.rs]: If an entity's `Name` or parent/children change, the path no longer is valid. #6785 introduces a per-path cache of entities to reduce the work needed for searching for the target entity. We could do something like that to reduce the load from the searches. That probably could work until the full animation composition RFC is merged and implemented, which has a more holistic solution to this.

Comment (smessmer) [reply crates/bevy_animation/src/lib.rs]: That works for me. We can merge this PR and then when #6785 gets merged, we'll automatically have the cache.

**Comment by smessmer** [comment]
Anything else to do before we can merge this?

**Comment by mockersf** [comment]
Should the new animation pushed keep the playback speed of the currently playing animation? This PR resets it. This was already the case, but I find it more disturbing now as there is a smooth transition instead of the break from before.

Perf wise, triggering a transition in `many_foxes` pretty much halves the FPS, from 110 to 65.

**Comment by smessmer** [comment]
@mockersf Did you change anything in the example? Running many_foxes as-is from this branch and from the version of the main branch it is based on gives me about the same fps.

**Comment by smessmer** [comment]
I ran an experiment adding a simple `HashMap<EntityPath, Entity>` that is cleared at the beginning of the frame but can be used when the same `EntityPath` is looked up multiple times. See https://github.com/smessmer/bevy/commit/f8af0f664b9fdaf4897e62419bef04a75ed08f73.

But that made things worse. As said before, I didn't see a significant frame rate difference for many_foxes between this PR and the main branch. But with the cache, it dropped to 50% fps for me. I don't fully understand why. Maybe hashing `EntityPath` is too expensive.

**Comment by smessmer** [comment]
About the animation speed, I made it part of the `PlayingAnimation` class. This means the speed of your previous animation will stay the same when you transition to a new animation. Only the new animation's speed will be reset. The reasoning for this is that you may have different animations that you want to run at different speeds. Say you have an "attack" animation that you want to run at 3x speed, and a "walking" animation that you want to run at either 1x or 2x speed depending on whether the character is walking or running. When transitioning between walking and attacking, you want the walking/running animation to stay walking/running and not change its speed while the transition is ongoing, but you want to set a new speed for the attacking animation.

**Comment by smessmer** [comment]
ping

**Review by james7132 — APPROVED** [review]
LGTM on my end. A lot of the perf issues should be addressed with #6785, or at least reeled back to manageable levels with it, as well as the subsequent RFCs.

**Comment by mockersf** [comment]
> @mockersf Did you change anything in the example? Running many_foxes as-is from this branch and from the version of the main branch it is based on gives me about the same fps. Tested both on a Laptop with Integrated Graphics and on a NVIDIA card. Although I guess this would be CPU bound rather...

I made the transition last longer, to have time to look at the FPS

**Comment by mockersf** [comment]
> LGTM on my end. A lot of the perf issues should be addressed with #6785, or at least reeled back to manageable levels with it, as well as the subsequent RFCs.

Would it be hard to create a branch with both those PR to check perfs?

**Comment by smessmer** [comment]
ok, I rebased it

**Review by mockersf — APPROVED** [review]
On many foxes, increasing transition from 250ms to 5s, I see the FPS drop from ~220 to ~175 during the transition.

Good enough for me, thanks!

**Review by alice-i-cecile — APPROVED** [review]
Great to have even basic animation blending in: this is essential for making use of animations.

bors r+

**Pull request closed without merging**
